### PR TITLE
[LA-46] [LA-48] Refactor cloudlrs-ingest-microservice code with Caliper support 

### PR DESCRIPTION
**Pull request by sandeepmjay**
Refactored xapi-ingest-microservice to lrs-ingest-microservice. 

The new service includes 
1) Support for both xapi and Caliper ingestion.
2) LRS data model changes on statements tables which includes actor/statement type. 
3) Caliper event validation
4) Custom Error handling response objects from Lambda handler which include Cloud Watch log stream information. This enables easier log tracking on the function invoker application (LRS poller).
5) Logging refactor to suite Lambda requirements.
6) Minor bug fixes on JSON parsing to extract targeted data from Caliper
7) Also changing the directory structure that will be compatible with AWS Code Pipeline.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+ * REGENTS SPECIFICALLY DISCLAIMS ANY WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE
+ * SOFTWARE AND ACCOMPANYING DOCUMENTATION, IF ANY, PROVIDED HEREUNDER IS PROVIDED
+ * "AS IS". REGENTS HAS NO OBLIGATION TO PROVIDE MAINTENANCE, SUPPORT, UPDATES,
+ * ENHANCEMENTS, OR MODIFICATIONS.
+ */
+
+var pg = require('pg');
+var config = require('config');
+var joi = require('joi');
```
Comment (johncrossman) [inline index.js]: @sandeepmjay convention tells us to use `Joi`, not `joi`. No biggie.

File: index.js [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+ * @param  {Object}         context                   Context object contains lambda runtime information such as functionName, CloudWatch log group etc.
+ * @param  {Object}         callback.err              An error that occurred with custom response payload, if any
+ * @param  {Object}         callback.response         Response object with the result and log stream information packaged as payload
+ */
+exports.handler = function(event, context, callback) {
+  require('./globals/sequelize.js');
+  var request = null;
+  var response = {
+      'logGroupName': context.logGroupName,
+      'logStreamName': context.logStreamName,
```
Comment (johncrossman) [inline index.js]: @sandeepmjay too much indentation here

File: index.js [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+      console.log("Request is not a well formed JSON");
+      response.result = 'Failed';
+      response.code = 400;
+      response.msg = 'Request not a well formed JSON';
+      context.callbackWaitsForEmptyEventLoop = false;
+      return callback(JSON.stringify(response));
+    }
+  } else if (typeof event === 'object') {
+    request = event;
+  }
```
Comment (johncrossman) [inline index.js]: @sandeepmjay what if event is neither 'string' nor 'object'?

File: index.js [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+                response.msg = 'Caliper statement processing successful with uuid: ' + statement.uuid;
+                context.callbackWaitsForEmptyEventLoop = false;
+                return callback(null, response);
+              }
+            });
+            break;
+          default:
+            console.log('Statement not in XAPI or Caliper format');
+            response.result = 'Failed';
+            response.code = err.code || 400;
```
Comment (johncrossman) [inline index.js]: @sandeepmjay same question (re: lines 124 and 142)

File: lib/caliper.js [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+  }
+
+  getStatement(statement.uuid, function(err, retrievedStatement) {
+    if (retrievedStatement) {
+      console.log({'id': statement.uuid}, 'Attempted to save a learning activity statement that already exists');
+      return callback({'code': 400, 'msg': 'Attempted to save a learning activity statement that already exists'});
+    } else if (err && err.code !== 404) {
+      console.log({'id': statement.uuid}, 'Unable to verify if learning activity statement already exists');
+      return callback(err);
+    }
```
Comment (johncrossman) [inline lib/caliper.js]: @sandeepmjay what if err.code is 404?

File: lib/caliper.js [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+};
+
+/**
+ * Check required Event properties against set of user-supplied values
+ * @param  {Object}           statement                 Caliper event to be validated
+ * @param  {Object}           callback.err              An error that occurred, if any
+ * @param  {Object}           callback.statement        The statement that passed validation
+ */
+var validateStatement = module.exports.validateStatement = function(statement, callback) {
+  var validationError = null ;
```
Comment (johncrossman) [inline lib/caliper.js]: @sandeepmjay extra space here

File: package.json [hunk 7]
```diff
@@ -1,10 +1,10 @@
 {
-  "name": "xapi-ingest-microservice",
-  "description": "XAPI ingestion service for the cloud-based Learning Record Store",
+  "name": "lrs-ingest-microservice",
+  "description": "LRS ingestion microservice for the cloud-based Learning Record Store to process XAPI and Caliper events",
   "version": "0.1.0",
   "repository": {
     "type": "git",
-    "url": "git://github.com/ets-berkeley-edu/cloudlrs-lambda-services.git"
+    "url": "git://github.berkeley.edu/ETS/cloudlrs-lambda-services.git"
```
Comment (johncrossman) [inline package.json]: @sandeepmjay can we rollback to github.com URL?  So confusing! :)

**Comment by johncrossman** [comment]
Phew! I'm done, @sandeepmjay  Let me know when amendments are made and I'll merge it.

File: index.js (hunk 1 above)
Comment (sandeepmjay) [inline index.js]: All of LRS code has this convention. Might need an overall refactor at some point. Benching this till then

**Comment by sandeepmjay** [comment]
@johncrossman Thanks for the review. Made amends based on comments. Go ahead with the merge !

**Pull request merged**